Kernel/Tasks/WaitQueue.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2025, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Badge.h>
+#include <Kernel/Arch/Processor.h>
+#include <Kernel/Tasks/Scheduler.h>
+#include <Kernel/Tasks/Thread.h>
+#include <Kernel/Tasks/WaitQueue.h>
+
+// Based on https://read.seas.harvard.edu/cs1610/2025/doc/wait-queues
+
+namespace Kernel {
+
+void WaitQueue::notify_all()
+{
+    SpinlockLocker queue_lock { m_lock };
+    while (!m_waiters.is_empty()) {
+        Waiter& waiter = *m_waiters.take_first();
+        waiter.notify({});
+    }
+}
+
+void WaitQueue::notify_one()
+{
+    SpinlockLocker queue_lock { m_lock };
+    if (m_waiters.is_empty())
+        return;
+
+    Waiter& waiter = *m_waiters.take_first();
+    waiter.notify({});
+}
+
+void Waiter::notify(Badge<WaitQueue>)
+{
+    VERIFY(m_association.has_value());
+    SpinlockLocker scheduler_lock(g_scheduler_lock);
+    auto& thread = *m_association->thread;
+
+    // If the thread is stopped, make the scheduler think that the
+    // thread was running beforehand so that the thread runs whenever
+    // it gets resumed (instead of remaining stopped/blocked forever).
+    if (thread.state() == Thread::State::Stopped)
+        thread.override_stop_state(Thread::State::Runnable);
```

Comment:
Now that I see how these `Thread::{override_stop_state,set_manual_block_handling}()` functions are used, I'm a bit skeptical.
I have never looked at the scheduler code, but this seems like a hack to make this work with the current scheduler that assumes blockers and mutexes.
I wonder if we could do this in a nicer way that doesn't require us to explicitly disable behavior. It would be nice if the default behavior (i.e. with no current blocker or mutex attached) of the scheduler would behave corretcly.
I'm also not sure if using these new functions triggers any edge cases.

Reply:
`set_manual_block_handling` makes sense to me. Though it would be nice to get rid of the current blocker infrastructure and have the scheduler not be involved like this.

I'm more curious why `override_stop_state` is needed. When would `set_state` set `m_stop_state` to an incorrect previous state?
I feel like there might be some edge cases where we don't want to return to Runnable.

Reply:
I'm just a bit confused what exact situation you are trying to prevent.
I assume this code is to make SIGSTOP work correctly while the thread is blocked, right? AFAICT threads can only be stopped with SIGSTOP.

But shouldn't a thread never be scheduled while it's blocked?
Signal delivery only happens while a thread is running, so I don't think a Blocked -> Stopped transition is possible.

Reply:
I'm honestly not really sure what we need. I think the best option might be to try one approach and see if it works (safely) in actual code.